src/lib/storage/segment_access_statistics.cpp

```diff
@@ -0,0 +1,171 @@
+#include "segment_access_statistics.hpp"
+
+namespace opossum {
+
+std::chrono::time_point<std::chrono::steady_clock> AtomicTimedAccessStrategy::start_time = std::chrono::steady_clock::now();
+std::chrono::duration<double, std::milli> AtomicTimedAccessStrategy::interval = std::chrono::milliseconds{1000};
+
+// -------------------------------------------------------------------------------------------------------------------
+SegmentAccessType SegmentAccessTypeTools::iterator_access_pattern(const std::shared_ptr<const PosList>& positions) {
+  const auto max_items_to_compare = std::min(positions->size(), 100ul);
+
+  auto state = State::Unknown;
+  for (auto i = 1ul; i < max_items_to_compare; ++i) {
+    const int64_t diff = static_cast<int64_t>(positions->operator[](i).chunk_offset) -
+                         static_cast<int64_t>(positions->operator[](i - 1).chunk_offset);
+
+    auto input = Input::Negative;
+    if (diff == 0) input = Input::Zero;
+    else if (diff == 1) input = Input::One;
+    else if (diff > 0) input = Input::Positive;
+    else if (diff == -1) input = Input::NegativeOne;
+
+    state = _transitions[static_cast<uint32_t>(state)][static_cast<uint32_t>(input)];
+  }
+
+  switch(state) {
+    case State::Unknown: case State::SeqInc: case State::SeqDec: return SegmentAccessType::IteratorSeqAccess;
+    case State::RndInc: case State::RndDec: return SegmentAccessType::IteratorIncreasingAccess;
+    case State::Rnd: return SegmentAccessType::IteratorRandomAccess;
+    case State::Count: Fail("Unknown state.");
+  }
+
+  Fail("This line should never be reached. It is required so the compiler won't complain.");
+}
+
+// -------------------------------------------------------------------------------------------------------------------
+uint64_t AtomicAccessStrategy::count(SegmentAccessType type) const {
```

Comment:
"Count" sounds like you are actively counting something (which is an expensive operation)

Reply:
I see your point. In this case sth. like get_count would be better.